Mathlib/Tactic/Backtracking.lean

```diff
@@ -0,0 +1,159 @@
+  /-- If `suspend g`, then we do not attempt to apply any further lemmas,
+  but return `g` as a new subgoal. (defaults to `false`) -/
+  suspend : MVarId → MetaM Bool := fun _ => pure false
+  /-- `discharge g` is called on goals for which no lemmas apply.
+  If `none` we return `g` as a new subgoal.
+  If `some l`, we replace `g` by `l` in the list of active goals, and recurse.
+  If failure, we backtrack. (defaults to failure) -/
+  discharge : MVarId → MetaM (Option (List MVarId)) := fun _ => failure
+
+/--
+Attempts to solve the `goals`, by recursively calling `alternatives g` on each subgoal that appears.
+`alternatives` returns a list of list of goals (wrapped in `MetaM`).
+The outermost list corresponds to alternative outcomes,
+while the innermost list is the subgoals generated in that outcome.
+
+`backtrack` performs a backtracking search, attempting to close all subgoals.
+
+Further flow control options are available via the `Config` argument.
+-/
+def backtrack (cfg : BacktrackConfig := {}) (trace : Name := .anonymous)
+    (alternatives : MVarId → MetaM (List (MetaM (List MVarId))))
+    (goals : List MVarId) : MetaM (List MVarId) := do
+run cfg.maxDepth goals []
+  where
+  /--
+  * `n : Nat` steps remaining.
+  * `curr : List MVarId` the current list of unsolved goals.
+  * `acc : List MVarId` a list of "suspended" goals, which will be returned as subgoals.
+  -/
+  run (n : Nat) (curr acc : List MVarId) : MetaM (List MVarId) := do
+  match n with
+  | 0 => do
+    -- We're out of fuel.
+    if cfg.failAtMaxDepth then
+      throwError "backtrack exceeded the recursion limit"
+    else
+      -- Before returning the goals, we run `cfg.proc` one last time.
+      let curr := acc.reverse ++ curr
+      return (← cfg.proc goals curr).getD curr
+  | n + 1 => do
+  -- First, run `cfg.proc`, to see if it wants to modify the goals.
+  match ← cfg.proc goals curr with
+  | some curr' => run n curr' acc
+  | none =>
+  match curr with
+  -- If there are no active goals, return the accumulated goals.
+  | [] => return acc.reverse
+  | g :: gs =>
+  -- Discard any goals which have already been assigned.
+  if ← g.isAssigned then
+    run (n+1) gs acc
+  else
+  withTraceNode trace
+    -- Note: the `addMessageContextFull` ensures we show the goal using the mvar context before
+    -- the `do` block below runs, potentially unifying mvars in the goal.
+    (return m!"{exceptEmoji ·} working on: {← addMessageContextFull g}")
+    do
+      -- Check if we should suspend the search here:
+      if (← cfg.suspend g) then
+        withTraceNode trace
+          (fun _ => return m!"⏸️ suspending search and returning as subgoal") do
+        run (n+1) gs (g :: acc)
+      else
+        try
+          -- We attempt to find an expression which can be applied,
+          -- and for which all resulting sub-goals can be discharged using `run n`.
+          g.firstContinuation alternatives (fun res => run n (res ++ gs) acc)
+        catch _ =>
```

Comment:
Should we trace the error in the `_`?

Reply:
There's no error here, it is just `firstContinuation` failing to find anything that works.